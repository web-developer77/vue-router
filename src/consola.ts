// Own simplified consola version
// Consola doesn't work on IE9

type LogCommand = 'info' | 'log' | 'error' | 'warn'
const logs: LogCommand[] = ['info', 'log', 'error', 'warn']

export default {
  tag: '',
  withTag(tag: string) {
    const cs = Object.create(this)
    cs.tag = tag
    return cs
  },

  info(...args: any[]) {
    if (this.tag) console.info(`[${this.tag}]`, ...args)
    else console.info(...args)
  },

  mockTypes(spyCreator: () => any): void {
    for (const log of logs) {
      // @ts-ignore
      this[log] = spyCreator()
    }
  },
}
