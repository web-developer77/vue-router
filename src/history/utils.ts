import {
  HistoryLocationNormalized,
  HistoryQuery,
  HistoryLocation,
} from './base'
import { RouteQuery } from '../types'

const PERCENT_RE = /%/g

/**
 * Transforms a URL into an object
 * @param location location to normalize
 * @param currentLocation current location, to reuse params and location
 */
export function parseURL(location: string): HistoryLocationNormalized {
  let path = '',
    query: HistoryQuery = {},
    searchString = '',
    hash = ''

  // Could use URL and URLSearchParams but IE 11 doesn't support it
  const searchPos = location.indexOf('?')
  const hashPos = location.indexOf('#', searchPos > -1 ? searchPos : 0)
  if (searchPos > -1) {
    path = location.slice(0, searchPos)
    searchString = location.slice(
      searchPos + 1,
      hashPos > -1 ? hashPos : location.length
    )

    query = parseQuery(searchString)
  }

  if (hashPos > -1) {
    path = path || location.slice(0, hashPos)
    hash = location.slice(hashPos, location.length)
  }

  path = path || location

  return {
    fullPath: location,
    path,
    query,
    hash,
  }
}

/**
 * Transform a queryString into a query object. Accept both, a version with the leading `?` and without
 * Should work as URLSearchParams
 * @param search
 */
export function parseQuery(search: string): HistoryQuery {
  // TODO: optimize by using a for loop
  const hasLeadingIM = search[0] === '?'
  return (hasLeadingIM ? search.slice(1) : search).split('&').reduce(
    (query: HistoryQuery, entry: string) => {
      const [key, value] = entry.split('=')
      if (key in query) {
        // an extra variable for ts types
        let currentValue = query[key]
        if (!Array.isArray(currentValue)) {
          currentValue = query[key] = [currentValue]
        }
        currentValue.push(value)
      } else {
        query[key] = value
      }

      return query
    },
    {} as HistoryQuery
  )
}

/**
 * Stringify a URL object
 * @param location
 */
export function stringifyURL(location: HistoryLocation): string {
  let url = location.path
  let query = location.query ? stringifyQuery(location.query) : ''

  return url + (query && '?' + query) + (location.hash || '')
}

/**
 * Stringify an object query. Works like URLSearchParams. Doesn't prepend a `?`
 * @param query
 */
export function stringifyQuery(query: HistoryQuery): string {
  let search = ''
  // TODO: util function?
  for (const key in query) {
    if (search.length > 1) search += '&'
    // TODO: handle array
    const value = query[key]
    if (Array.isArray(value)) {
      search += `${key}=${value[0]}`
      for (let i = 1; i < value.length; i++) {
        search += `&${key}=${value[i]}`
      }
    } else {
      search += `${key}=${query[key]}`
    }
  }

  return search
}

export function normalizeQuery(query: RouteQuery): HistoryQuery {
  // TODO: implem
  return query as HistoryQuery
}

/**
 * Prepare a URI string to be passed to pushState
 * @param uri
 */
export function prepareURI(uri: string) {
  // encode the % symbol so it also works on IE
  return uri.replace(PERCENT_RE, '%25')
}

// use regular decodeURI
export const decodeURI = global.decodeURI

/**
 * Normalize a History location into an object that looks like
 * the one at window.location
 * @param location
 */
export function normalizeLocation(
  location: string | HistoryLocation
): HistoryLocationNormalized {
  if (typeof location === 'string') return parseURL(location)
  else
    return {
      fullPath: stringifyURL(location),
      path: location.path,
      query: location.query || {},
      hash: location.hash || '',
    }
}
