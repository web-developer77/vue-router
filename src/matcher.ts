import pathToRegexp from 'path-to-regexp'
import {
  RouteRecord,
  RouteParams,
  MatcherLocation,
  MatcherLocationNormalized,
  MatcherLocationRedirect,
  // TODO: add it to matched
  // MatchedRouteRecord,
} from './types/index'
import { NoRouteMatchError, InvalidRouteMatch } from './errors'

type NormalizedRouteRecord = Exclude<RouteRecord, { component: any }> // normalize component/components into components

interface RouteMatcher {
  re: RegExp
  resolve: (params?: RouteParams) => string
  record: NormalizedRouteRecord
  parent: RouteMatcher | void
  // TODO: children so they can be removed
  // children: RouteMatcher[]
  keys: string[]
  score: number
}

/**
 * Normalizes a RouteRecord into a MatchedRouteRecord. Creates a copy
 * @param record
 * @returns the normalized version
 */
export function normalizeRecord(
  record: Readonly<RouteRecord>
): NormalizedRouteRecord {
  if ('component' in record) {
    const { component, ...rest } = record
    // @ts-ignore I could do it type safe by copying again rest:
    // return {
    //   ...rest,
    //   components: { default: component }
    // }
    // but it's slower
    rest.components = { default: component }
    return rest as NormalizedRouteRecord
  }

  // otherwise just create a copy
  return { ...record }
}

enum PathScore {
  Segment = 4, // /a-segment
  SubSegment = 2, // /multiple-:things-in-one-:segment
  Static = 3, // /static
  Dynamic = 2, // /:someId
  DynamicCustomRegexp = 2.5, // /:someId(\\d+)
  Wildcard = -1, // /:namedWildcard(.*)
  SubWildcard = 1, // Wildcard as a subsegment
  Repeatable = -0.5, // /:w+ or /:w*
  Optional = -4, // /:w? or /:w*
  SubOptional = -0.1, // optional inside a subsegment /a-:w? or /a-:w*
  Root = 1, // just /
}

// allows to check if the user provided a custom regexp
const isDefaultPathRegExpRE = /^\[\^[^\]]+\]\+\?$/

export function createRouteMatcher(
  record: Readonly<NormalizedRouteRecord>,
  parent: RouteMatcher | void,
  options: pathToRegexp.RegExpOptions
): RouteMatcher {
  const keys: pathToRegexp.Key[] = []
  // options only use `delimiter`
  const tokens = pathToRegexp.parse(record.path, options)
  const re = pathToRegexp.tokensToRegExp(tokens, keys, options)
  // we pass a copy because later on we are modifying the original token array
  // to compute the score of routes
  const resolve = pathToRegexp.tokensToFunction([...tokens])

  let score = 0

  // console.log(tokens)
  // console.log('--- GROUPING ---')

  // special case for root path
  if (tokens.length === 1 && tokens[0] === '/') {
    score = 5
  } else {
    // allows us to group tokens into one single segment
    // it will point to the first token of the current group
    let currentSegment = 0
    // we group them in arrays to process them later
    const groups: Array<pathToRegexp.Token[]> = []
    // we skip the first element as it must be part of the first group
    const token = tokens[0]
    if (typeof token === 'string') {
      // TODO: refactor code in loop (right now it is duplicated)
      // we still need to check for / inside the string
      // remove the empty string because of the leading slash
      const sections = token.split('/').slice(1)
      if (sections.length > 1) {
        // the last one is going to replace currentSegment
        const last = sections.pop() as string // ts complains but length >= 2
        // we need to finalize previous group but we cannot use current entry
        // here we are sure that currentSegment < i because the token doesn't start with /
        // assert(currentSegment < i)
        const first = sections.shift() as string // ts complains but length >= 2
        // so we cut until the current segment and add the first section of current token as well
        groups.push(tokens.slice(currentSegment, 0).concat('/' + first))
        // equivalent to
        // groups.push(['/' + first])

        // we add the remaining sections as static groups
        for (const section of sections) {
          groups.push(['/' + section])
        }

        // we replace current entry with the last section and reset current segment
        tokens[0] = '/' + last
        // currentSegment = 0
      }
    }

    for (let i = 1; i < tokens.length; i++) {
      const token = tokens[i]
      if (typeof token === 'string') {
        if (token.charAt(0) === '/') {
          // finalize previous group and start a new one
          groups.push(tokens.slice(currentSegment, i))
          currentSegment = i
        } else {
          // we still need to check for / inside the string
          const sections = token.split('/')
          if (sections.length > 1) {
            // the last one is going to replace currentSegment
            const last = sections.pop() as string // ts complains but length >= 2
            // we need to finalize previous group but we cannot use current entry
            // here we are sure that currentSegment < i because the token doesn't start with /
            // assert(currentSegment < i)
            const first = sections.shift() as string // ts complains but length >= 2
            // so we cut until the current segment and add the first section of current token as well
            groups.push(tokens.slice(currentSegment, i).concat(first))

            // we add the remaining sections as static groups
            for (const section of sections) {
              groups.push(['/' + section])
            }

            // we replace current entry with the last section and reset current segment
            tokens[i] = '/' + last
            currentSegment = i
          }
        }
      } else {
        if (token.prefix.charAt(0) === '/') {
          // finalize previous group and start a new one
          groups.push(tokens.slice(currentSegment, i))
          currentSegment = i
        }
      }
    }

    // add the remaining segment as one group
    // TODO: refactor the handling of ending with static like /:a/b/c
    if (currentSegment < tokens.length) {
      let token: pathToRegexp.Token
      if (
        tokens.length - currentSegment === 1 &&
        typeof (token = tokens[tokens.length - 1]) === 'string'
      ) {
        // the remaining group is a single string, so it must start with a leading /
        const sections = token.split('/').slice(1)
        // we add the remaining sections as static groups
        for (const section of sections) {
          groups.push(['/' + section])
        }
      } else {
        groups.push(tokens.slice(currentSegment))
      }
    }

    const scoreForSegment = function scoreForSegment(
      group: pathToRegexp.Token
    ): number {
      let score = PathScore.Segment
      if (typeof group === 'string') {
        score += group === '/' ? PathScore.Root : PathScore.Static
      } else {
        score +=
          group.pattern === '.*'
            ? PathScore.Wildcard
            : isDefaultPathRegExpRE.test(group.pattern)
            ? PathScore.Dynamic
            : PathScore.DynamicCustomRegexp
        score +=
          +group.optional * PathScore.Optional +
          +group.repeat * PathScore.Repeatable
        if (typeof group.name === 'number') throw new Error('Name your param')
        // keys.push(group.name)
      }
      return score
    }

    const scoreForSubSegment = function scoreForSubSegment(
      group: pathToRegexp.Token
    ): number {
      let score = 0
      if (typeof group === 'string') {
        // in a sub segment, it doesn't matter if it's root or not
        score += PathScore.Static
      } else {
        score +=
          group.pattern === '.*'
            ? PathScore.SubWildcard
            : isDefaultPathRegExpRE.test(group.pattern)
            ? PathScore.Dynamic
            : PathScore.DynamicCustomRegexp
        score += +group.optional * PathScore.SubOptional
        if (typeof group.name === 'number') throw new Error('Name your param')
        // keys.push(group.name)
      }
      return score
    }

    for (const group of groups) {
      // console.log(group)
      if (group.length === 1) {
        score += scoreForSegment(group[0])
      } else {
        score += PathScore.Segment + PathScore.SubSegment
        let multiplier = 1 / 10
        for (let i = 0; i < group.length; i++) {
          score += scoreForSubSegment(group[i]) * multiplier
          multiplier /= 10
        }
      }
      // segments.push('/' + section)
    }

    // console.log(record.path, { score })
    // console.log('____'.repeat(20))
  }

  // create the object before hand so it can be passed to children
  return {
    parent,
    record: record,
    re,
    // TODO: handle numbers differently. Maybe take the max one and say there are x unnamed keys
    keys: keys.map(key => String(key.name)),
    resolve,
    score,
  }
}

export class RouterMatcher {
  private matchers: RouteMatcher[] = []

  constructor(routes: RouteRecord[]) {
    for (const route of routes) {
      this.addRouteRecord(route)
    }
  }

  private addRouteRecord(
    record: Readonly<RouteRecord>,
    parent?: RouteMatcher
  ): void {
    const options: pathToRegexp.RegExpOptions = {}

    const recordCopy = normalizeRecord(record)

    if (parent) {
      // if the child isn't an absolute route
      if (record.path[0] !== '/') {
        let path = parent.record.path
        // only add the / delimiter if the child path isn't empty
        if (recordCopy.path) path += '/'
        path += record.path
        recordCopy.path = path
      }
    }

    // create the object before hand so it can be passed to children
    const matcher = createRouteMatcher(recordCopy, parent, options)

    if ('children' in record && record.children) {
      for (const childRecord of record.children) {
        this.addRouteRecord(childRecord, matcher)
      }
    }

    this.insertMatcher(matcher)
  }

  private insertMatcher(matcher: RouteMatcher) {
    let i = 0
    while (i < this.matchers.length && matcher.score <= this.matchers[i].score)
      i++
    this.matchers.splice(i, 0, matcher)
  }

  /**
   * Resolve a location without doing redirections so it can be used for anchors
   */
  resolveAsPath() {}

  /**
   * Transforms a MatcherLocation object into a normalized location
   * @param location MatcherLocation to resolve to a url
   */
  resolve(
    location: Readonly<MatcherLocation>,
    currentLocation: Readonly<MatcherLocationNormalized>
  ): MatcherLocationNormalized | MatcherLocationRedirect {
    let matcher: RouteMatcher | void
    let params: RouteParams = {}
    let path: MatcherLocationNormalized['path']
    let name: MatcherLocationNormalized['name']

    if ('path' in location) {
      matcher = this.matchers.find(m => m.re.test(location.path))

      // TODO: if no matcher, return the location with an empty matched array
      // to allow non existent matches
      // TODO: warning of unused params if provided
      if (!matcher) throw new NoRouteMatchError(currentLocation, location)

      // no need to resolve the path with the matcher as it was provided
      path = location.path
      name = matcher.record.name

      // fill params
      const result = matcher.re.exec(path)

      if (!result) {
        throw new Error(`Error parsing path "${location.path}"`)
      }

      for (let i = 0; i < matcher.keys.length; i++) {
        const key = matcher.keys[i]
        const value = result[i + 1]
        if (!value) {
          throw new Error(
            `Error parsing path "${location.path}" when looking for param "${key}"`
          )
        }
        params[key] = value
      }

      if ('redirect' in matcher.record) {
        const { redirect } = matcher.record
        return {
          redirect,
          normalizedLocation: {
            name,
            path,
            matched: [],
            params,
          },
        }
      }
      // named route
    } else if ('name' in location) {
      matcher = this.matchers.find(m => m.record.name === location.name)

      if (!matcher) throw new NoRouteMatchError(currentLocation, location)

      name = matcher.record.name
      params = location.params || currentLocation.params // TODO: normalize params
      path = matcher.resolve(params)
      // TODO: check missing params

      if ('redirect' in matcher.record) {
        const { redirect } = matcher.record
        return {
          redirect,
          normalizedLocation: {
            name,
            path,
            matched: [],
            params,
          },
        }
      }
      // location is a relative path
    } else {
      // match by name or path of current route
      matcher = currentLocation.name
        ? this.matchers.find(m => m.record.name === currentLocation.name)
        : this.matchers.find(m => m.re.test(currentLocation.path))
      if (!matcher) throw new NoRouteMatchError(currentLocation, location)
      name = matcher.record.name
      params = location.params || currentLocation.params
      path = matcher.resolve(params)
    }

    // this should never happen because it will mean that the user ended up in a route
    // that redirects but ended up not redirecting
    if ('redirect' in matcher.record) throw new InvalidRouteMatch(location)

    const matched: MatcherLocationNormalized['matched'] = [matcher.record]
    let parentMatcher: RouteMatcher | void = matcher.parent
    while (parentMatcher) {
      // reversed order so parents are at the beginning
      // TODO: should be doable by typing RouteMatcher in a different way
      if ('redirect' in parentMatcher.record) throw new Error('TODO')
      matched.unshift(parentMatcher.record)
      parentMatcher = parentMatcher.parent
    }

    return {
      name,
      path,
      params,
      matched,
    }
  }
}
