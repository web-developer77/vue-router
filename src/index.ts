import { Router } from './router'
import { HTML5History } from './history/html5'

export { Router, HTML5History }
