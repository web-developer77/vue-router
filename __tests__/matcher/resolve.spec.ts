import { RouterMatcher } from '../../src/matcher'
import {
  START_LOCATION_NORMALIZED,
  RouteComponent,
  RouteRecord,
  MatcherLocation,
  MatcherLocationNormalized,
  MatcherLocationRedirect,
} from '../../src/types'
import { normalizeRouteRecord } from '../utils'

// @ts-ignore
const component: RouteComponent = null

function createRouterMatcher(records: RouteRecord[]) {
  return new RouterMatcher(records)
}

// for normalized records
const components = { default: component }

describe('Router Matcher', () => {
  describe('resolve', () => {
    function assertRecordMatch(
      record: RouteRecord | RouteRecord[],
      location: MatcherLocation,
      resolved: Partial<MatcherLocationNormalized>,
      start: MatcherLocationNormalized = START_LOCATION_NORMALIZED
    ) {
      record = Array.isArray(record) ? record : [record]
      const matcher = createRouterMatcher(record)

      if (!('meta' in resolved)) {
        resolved.meta = record[0].meta || {}
      }

      // add location if provided as it should be the same value
      if ('path' in location && !('path' in resolved)) {
        resolved.path = location.path
      }

      if ('redirect' in record) {
        throw new Error('not handled')
      } else {
        // use one single record
        if (!resolved.matched)
          // @ts-ignore
          resolved.matched = record.map(normalizeRouteRecord)
        else resolved.matched = resolved.matched.map(normalizeRouteRecord)
      }

      // allows not passing params
      if ('params' in location) {
        resolved.params = resolved.params || location.params
      } else {
        resolved.params = resolved.params || {}
      }

      const startCopy = {
        ...start,
        matched: start.matched.map(normalizeRouteRecord),
      }

      // make matched non enumerable
      Object.defineProperty(startCopy, 'matched', { enumerable: false })

      const result = matcher.resolve(location, startCopy)
      expect(result).toEqual(resolved)
    }

    /**
     *
     * @param record Record or records we are testing the matcher against
     * @param location location we want to reolve against
     * @param [start] Optional currentLocation used when resolving
     * @returns error
     */
    function assertErrorMatch(
      record: RouteRecord | RouteRecord[],
      location: MatcherLocation,
      start: MatcherLocationNormalized = START_LOCATION_NORMALIZED
    ): any {
      try {
        assertRecordMatch(record, location, {}, start)
      } catch (error) {
        return error
      }
    }

    describe('alias', () => {
      it('resolves an alias', () => {
        assertRecordMatch(
          {
            path: '/',
            alias: '/home',
            name: 'Home',
            components,
            meta: { foo: true },
          },
          { path: '/home' },
          {
            name: 'Home',
            path: '/home',
            params: {},
            meta: { foo: true },
            matched: [
              {
                path: '/home',
                name: 'Home',
                components,
                meta: { foo: true },
              },
            ],
          }
        )
      })

      it('resolves an alias with children', () => {
        const children = [{ path: 'one', component, name: 'nested' }]
        assertRecordMatch(
          {
            path: '/parent',
            alias: '/p',
            component,
            children,
          },
          { path: '/p/one' },
          {
            path: '/p/one',
            name: 'nested',
            params: {},
            matched: [
              { path: '/p', children, components },
              { path: '/p/one', name: 'nested', components },
            ],
          }
        )
      })
    })

    describe('LocationAsPath', () => {
      it('resolves a normal path', () => {
        assertRecordMatch(
          { path: '/', name: 'Home', components },
          { path: '/' },
          { name: 'Home', path: '/', params: {} }
        )
      })

      it('resolves a normal path without name', () => {
        assertRecordMatch(
          { path: '/', components },
          { path: '/' },
          { name: undefined, path: '/', params: {} }
        )
      })

      it('resolves a path with params', () => {
        assertRecordMatch(
          { path: '/users/:id', name: 'User', components },
          { path: '/users/posva' },
          { name: 'User', params: { id: 'posva' } }
        )
      })

      it('resolves an array of params for a repeatable params', () => {
        assertRecordMatch(
          { path: '/a/:p+', name: 'a', components },
          { name: 'a', params: { p: ['b', 'c', 'd'] } },
          { name: 'a', path: '/a/b/c/d', params: { p: ['b', 'c', 'd'] } }
        )
      })

      it('resolves single params for a repeatable params', () => {
        assertRecordMatch(
          { path: '/a/:p+', name: 'a', components },
          { name: 'a', params: { p: 'b' } },
          { name: 'a', path: '/a/b', params: { p: 'b' } }
        )
      })

      it('keeps repeated params as a single one when provided through path', () => {
        assertRecordMatch(
          { path: '/a/:p+', name: 'a', components },
          { path: '/a/b/c' },
          // TODO: maybe it should consistently be an array for repeated params
          { name: 'a', params: { p: 'b/c' } }
        )
      })

      it('resolves a path with multiple params', () => {
        assertRecordMatch(
          { path: '/users/:id/:other', name: 'User', components },
          { path: '/users/posva/hey' },
          { name: 'User', params: { id: 'posva', other: 'hey' } }
        )
      })

      it('resolves a path with multiple params but no name', () => {
        assertRecordMatch(
          { path: '/users/:id/:other', components },
          { path: '/users/posva/hey' },
          { name: undefined, params: { id: 'posva', other: 'hey' } }
        )
      })

      it('throws if the path does not exists', () => {
        expect(
          assertErrorMatch({ path: '/', components }, { path: '/foo' })
        ).toMatchSnapshot()
      })
    })

    describe('LocationAsName', () => {
      it('matches a name', () => {
        assertRecordMatch(
          { path: '/home', name: 'Home', components },
          { name: 'Home' },
          { name: 'Home', path: '/home' }
        )
      })

      it('matches a name and fill params', () => {
        assertRecordMatch(
          { path: '/users/:id/m/:role', name: 'UserEdit', components },
          { name: 'UserEdit', params: { id: 'posva', role: 'admin' } },
          { name: 'UserEdit', path: '/users/posva/m/admin' }
        )
      })

      it('throws if the named route does not exists', () => {
        expect(
          assertErrorMatch({ path: '/', components }, { name: 'Home' })
        ).toMatchSnapshot()
      })
    })

    describe('LocationAsRelative', () => {
      it('matches with nothing', () => {
        const record = { path: '/home', name: 'Home', components }
        assertRecordMatch(
          record,
          {},
          { name: 'Home', path: '/home' },
          {
            name: 'Home',
            params: {},
            path: '/home',
            matched: [record],
            meta: {},
          }
        )
      })

      it('replace params even with no name', () => {
        const record = { path: '/users/:id/m/:role', components }
        assertRecordMatch(
          record,
          { params: { id: 'posva', role: 'admin' } },
          { name: undefined, path: '/users/posva/m/admin' },
          {
            path: '/users/ed/m/user',
            name: undefined,
            params: { id: 'ed', role: 'user' },
            matched: [record],
            meta: {},
          }
        )
      })

      it('replace params', () => {
        const record = {
          path: '/users/:id/m/:role',
          name: 'UserEdit',
          components,
        }
        assertRecordMatch(
          record,
          { params: { id: 'posva', role: 'admin' } },
          { name: 'UserEdit', path: '/users/posva/m/admin' },
          {
            path: '/users/ed/m/user',
            name: 'UserEdit',
            params: { id: 'ed', role: 'user' },
            matched: [],
            meta: {},
          }
        )
      })

      it('keep params if not provided', () => {
        const record = {
          path: '/users/:id/m/:role',
          name: 'UserEdit',
          components,
        }
        assertRecordMatch(
          record,
          {},
          {
            name: 'UserEdit',
            path: '/users/ed/m/user',
            params: { id: 'ed', role: 'user' },
          },
          {
            path: '/users/ed/m/user',
            name: 'UserEdit',
            params: { id: 'ed', role: 'user' },
            matched: [record],
            meta: {},
          }
        )
      })

      it('keep params if not provided even with no name', () => {
        const record = { path: '/users/:id/m/:role', components }
        assertRecordMatch(
          record,
          {},
          {
            name: undefined,
            path: '/users/ed/m/user',
            params: { id: 'ed', role: 'user' },
          },
          {
            path: '/users/ed/m/user',
            name: undefined,
            params: { id: 'ed', role: 'user' },
            matched: [record],
            meta: {},
          }
        )
      })

      describe('redirects', () => {
        function assertRedirect(
          records: RouteRecord[],
          location: MatcherLocation,
          expected: MatcherLocationNormalized | MatcherLocationRedirect,
          currentLocation: MatcherLocationNormalized = START_LOCATION_NORMALIZED
        ) {
          const matcher = createRouterMatcher(records)
          const resolved = matcher.resolve(location, currentLocation)
          expect(resolved).toEqual(expected)
          return resolved
        }

        it('resolves a redirect string', () => {
          const records = [
            { path: '/home', components },
            { path: '/redirect', redirect: '/home' },
          ]
          assertRedirect(
            records,
            {
              name: undefined,
              path: '/redirect',
            },
            {
              redirect: '/home',
              normalizedLocation: {
                path: '/redirect',
                params: {},
                name: undefined,
                matched: [],
                meta: {},
              },
            }
          )
        })

        it('resolves a redirect function that returns a string', () => {
          const redirect = () => '/home'
          const records = [
            { path: '/home', components },
            { path: '/redirect', redirect },
          ]
          assertRedirect(
            records,
            {
              name: undefined,
              path: '/redirect',
            },
            {
              redirect,
              normalizedLocation: {
                path: '/redirect',
                params: {},
                name: undefined,
                matched: [],
                meta: {},
              },
            }
          )
        })

        it('resolves a redirect function that returns an object route', () => {
          const redirect = () => {
            path: '/home'
          }
          const records = [
            { path: '/home', components },
            { path: '/redirect', redirect },
          ]
          assertRedirect(
            records,
            {
              name: undefined,
              path: '/redirect',
            },
            {
              redirect,
              normalizedLocation: {
                path: '/redirect',
                params: {},
                name: undefined,
                matched: [],
                meta: {},
              },
            }
          )
        })

        it('resolves a redirect as an object', () => {
          const records = [
            { path: '/home', components },
            { path: '/redirect', redirect: { path: 'home' } },
          ]
          assertRedirect(
            records,
            {
              name: undefined,
              path: '/redirect',
            },
            {
              redirect: { path: 'home' },
              normalizedLocation: {
                path: '/redirect',
                params: {},
                name: undefined,
                matched: [],
                meta: {},
              },
            }
          )
        })

        it('works with a named location', () => {
          const records = [
            { path: '/home', components },
            { path: '/redirect', name: 'redirect', redirect: { path: 'home' } },
          ]
          assertRedirect(
            records,
            {
              name: 'redirect',
            },
            {
              redirect: { path: 'home' },
              normalizedLocation: {
                path: '/redirect',
                params: {},
                name: 'redirect',
                matched: [],
                meta: {},
              },
            }
          )
        })

        it('throws if relative location when redirecting', () => {
          expect(
            assertErrorMatch(
              { path: '/redirect', redirect: '/home' },
              { params: {} },
              {
                path: '/redirect',
                params: {},
                matched: [],
                name: undefined,
                meta: {},
              }
            )
          ).toMatchSnapshot()
        })

        it('normalize a location when redirecting', () => {
          const redirect = (to: any) => ({ name: 'b', params: to.params })
          const records = [
            { path: '/home', components },
            {
              path: '/a/:a',
              name: 'a',
              redirect,
            },
            { path: '/b/:a', name: 'b', components },
          ]
          assertRedirect(
            records,
            {
              path: '/a/foo',
            },
            {
              redirect,
              normalizedLocation: {
                path: '/a/foo',
                params: { a: 'foo' },
                name: 'a',
                matched: [],
                meta: {},
              },
            }
          )
        })
      })

      it('throws if the current named route does not exists', () => {
        const record = { path: '/', components }
        const start = {
          name: 'home',
          params: {},
          path: '/',
          matched: [record],
        }
        // the property should be non enumerable
        Object.defineProperty(start, 'matched', { enumerable: false })
        expect(
          assertErrorMatch(
            record,
            {},
            {
              ...start,
              matched: start.matched.map(normalizeRouteRecord),
              meta: {},
            }
          )
        ).toMatchSnapshot()
      })
    })

    describe('children', () => {
      const ChildA = { path: 'a', name: 'child-a', components }
      const ChildB = { path: 'b', name: 'child-b', components }
      const ChildC = { path: 'c', name: 'child-c', components }
      const ChildWithParam = { path: ':p', name: 'child-params', components }
      const NestedChildWithParam = {
        ...ChildWithParam,
        name: 'nested-child-params',
      }
      const NestedChildA = { ...ChildA, name: 'nested-child-a' }
      const NestedChildB = { ...ChildB, name: 'nested-child-b' }
      const NestedChildC = { ...ChildC, name: 'nested-child-c' }
      const Nested = {
        path: 'nested',
        name: 'nested',
        components,
        children: [NestedChildA, NestedChildB, NestedChildC],
      }
      const NestedWithParam = {
        path: 'nested/:n',
        name: 'nested',
        components,
        children: [NestedChildWithParam],
      }

      it('resolves children', () => {
        const Foo = {
          path: '/foo',
          name: 'Foo',
          components,
          children: [ChildA, ChildB, ChildC],
        }
        assertRecordMatch(
          Foo,
          { path: '/foo/b' },
          {
            name: 'child-b',
            path: '/foo/b',
            params: {},
            matched: [Foo, { ...ChildB, path: `${Foo.path}/${ChildB.path}` }],
          }
        )
      })

      it('resolves children with empty paths', () => {
        const Nested = { path: '', name: 'nested', components }
        const Foo = {
          path: '/foo',
          name: 'Foo',
          components,
          children: [Nested],
        }
        assertRecordMatch(
          Foo,
          { path: '/foo' },
          {
            name: 'nested',
            path: '/foo',
            params: {},
            matched: [Foo, { ...Nested, path: `${Foo.path}` }].map(
              normalizeRouteRecord
            ),
          }
        )
      })

      it('resolves nested children with empty paths', () => {
        const NestedNested = { path: '', name: 'nested', components }
        const Nested = {
          path: '',
          name: 'nested-nested',
          components,
          children: [NestedNested],
        }
        const Foo = {
          path: '/foo',
          name: 'Foo',
          components,
          children: [Nested],
        }
        assertRecordMatch(
          Foo,
          { path: '/foo' },
          {
            name: 'nested',
            path: '/foo',
            params: {},
            matched: [
              Foo,
              { ...Nested, path: `${Foo.path}` },
              { ...NestedNested, path: `${Foo.path}` },
            ],
          }
        )
      })

      it('resolves nested children', () => {
        const Foo = {
          path: '/foo',
          name: 'Foo',
          components,
          children: [Nested],
        }
        assertRecordMatch(
          Foo,
          { path: '/foo/nested/a' },
          {
            name: 'nested-child-a',
            path: '/foo/nested/a',
            params: {},
            matched: [
              Foo,
              { ...Nested, path: `${Foo.path}/${Nested.path}` },
              {
                ...NestedChildA,
                path: `${Foo.path}/${Nested.path}/${NestedChildA.path}`,
              },
            ],
          }
        )
      })

      it('resolves nested children with named location', () => {
        const Foo = {
          path: '/foo',
          name: 'Foo',
          components,
          children: [Nested],
        }
        assertRecordMatch(
          Foo,
          { name: 'nested-child-a' },
          {
            name: 'nested-child-a',
            path: '/foo/nested/a',
            params: {},
            matched: [
              Foo,
              { ...Nested, path: `${Foo.path}/${Nested.path}` },
              {
                ...NestedChildA,
                path: `${Foo.path}/${Nested.path}/${NestedChildA.path}`,
              },
            ],
          }
        )
      })

      it('resolves nested children with relative location', () => {
        const Foo = {
          path: '/foo',
          name: 'Foo',
          components,
          children: [Nested],
        }
        assertRecordMatch(
          Foo,
          {},
          {
            name: 'nested-child-a',
            path: '/foo/nested/a',
            params: {},
            matched: [
              Foo,
              { ...Nested, path: `${Foo.path}/${Nested.path}` },
              {
                ...NestedChildA,
                path: `${Foo.path}/${Nested.path}/${NestedChildA.path}`,
              },
            ],
          },
          {
            name: 'nested-child-a',
            matched: [],
            params: {},
            path: '/foo/nested/a',
            meta: {},
          }
        )
      })

      it('resolves nested children with params', () => {
        const Foo = {
          path: '/foo',
          name: 'Foo',
          components,
          children: [NestedWithParam],
        }
        assertRecordMatch(
          Foo,
          { path: '/foo/nested/a/b' },
          {
            name: 'nested-child-params',
            path: '/foo/nested/a/b',
            params: { p: 'b', n: 'a' },
            matched: [
              Foo,
              {
                ...NestedWithParam,
                path: `${Foo.path}/${NestedWithParam.path}`,
              },
              {
                ...NestedChildWithParam,
                path: `${Foo.path}/${NestedWithParam.path}/${NestedChildWithParam.path}`,
              },
            ],
          }
        )
      })

      it('resolves nested children with params with named location', () => {
        const Foo = {
          path: '/foo',
          name: 'Foo',
          components,
          children: [NestedWithParam],
        }
        assertRecordMatch(
          Foo,
          { name: 'nested-child-params', params: { p: 'a', n: 'b' } },
          {
            name: 'nested-child-params',
            path: '/foo/nested/b/a',
            params: { p: 'a', n: 'b' },
            matched: [
              Foo,
              {
                ...NestedWithParam,
                path: `${Foo.path}/${NestedWithParam.path}`,
              },
              {
                ...NestedChildWithParam,
                path: `${Foo.path}/${NestedWithParam.path}/${NestedChildWithParam.path}`,
              },
            ],
          }
        )
      })
    })
  })
})
