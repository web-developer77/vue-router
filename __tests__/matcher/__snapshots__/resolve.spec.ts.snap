// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouterMatcher.resolve LocationAsName throws if the named route does not exists 1`] = `
[Error: No match for
 {"name":"Home"}]
`;

exports[`RouterMatcher.resolve LocationAsRelative throws if the current named route does not exists 1`] = `
[Error: No match for
 {"params":{"a":"foo"}}
while being at
{"name":"home","params":{},"path":"/","meta":{}}]
`;
